Retain prioritization information for closed streams

deweerdt opened the pull request:
This PR adds logic that has H2O record the priority information of the last 10 streams that were closed. 

This plays better with Chrome's chain of exclusive dependencies that might be re-ordered, because it makes it more likely that we'll find a stream in the even of a re-ordering. I've tested the code against https://github.com/pmeenan/http2priorities and that gave consistently good results in webpagetest (100 successive good runs).

kazuho commented:
To confirm, you succeeded i. reproducing the issue _and_ the issue when away with this change?

Sounds fantastic to me.

deweerdt commented:
> To confirm, you succeeded in reproducing the issue and the issue when away with this change?

That's right, here's a run with master that shows the issue before the fix:
https://www.webpagetest.org/result/181204_PW_2c05a1efceba6ef5771b28db79fa79e2/9/details/#waterfall_view_step1

I then ran 10 times the test which runs 9 page loads, and none of those had an issue.

kazuho commented on include/h2o/http2_internal.h:
```diff
@@ -170,6 +170,14 @@ struct st_h2o_http2_conn_t {
+    struct {
+        struct {
+            h2o_http2_scheduler_openref_t *sched_node;
+            uint32_t stream_id;
```
I notice that in `get_parent_stream_id` function of lib/http2/connection.c, we might refer to `h2o_http2_stream_t::stream_id` that is relative to the address of `sched_node`.

Considering that, it might make more sense to simply keep a pointer to `h2o_http2_stream_t` that has all the values disposed of with the exception being `stream_id` and `_refs.scheduler`, rather than retaining a pointer to `sched_node` as well as separately keeping `stream_id`.

Note however that we also agree that it's a bit of waste to retain `h2o_http2_stream_t` even after the stream is closed. FWIW, I think we could use something like 94536be to move just the scheduling node to a different location.

The way of adjusting this structure depends on if and how we might fix the issue.

deweerdt commented:
Thank you for writing `h2o_http2_scheduler_relocate`. It does simplify the code significantly. I believe e56120f addresses all your comments so far.
It introduces a `h2o_http2_closed_stream_t` struct and adds a `get_stream_id` callback in order to get the stream from a `h2o_http2_scheduler_node_t`

deweerdt commented:
@kazuho merged deweerdt#2, thanks!
